app/handlers/gate.go

```diff
@@ -62,27 +61,21 @@ func NewGateHandler(timeoutConfig int64, rulesConfigValue string, adapters []ada
 
 func (g gateHandler) PostOpen(c *gin.Context) {
 	ipNet := g.getIpNetFromContext(c)
-	rules := make([]domain.Rule, len(g.defaultRules))
-	copy(rules, g.defaultRules)
-	var adapterErrors []error
+	rules := g.createRules(ipNet)
 
+	errorDetails := make(map[string]string)
 	for _, adapter := range g.adapters {
-		for _, rule := range rules {
-			rule.IPNet = ipNet
-			err := g.callAdapter(adapter, rule)
-			if err != nil {
-				adapterErrors = append(adapterErrors, err)
-			}
+		callResult := g.callAdapter(adapter, rules)
+		if !callResult.IsSuccessful() {
```

Comment:
callAdapter also checks for `isSuccessful` consider implementing this differently to prevent confusion and stimulate single responsibility.